libmamba/include/mamba/core/graph_util.hpp

```diff
@@ -7,303 +7,493 @@
 #ifndef MAMBA_CORE_GRAPH_UTIL_HPP
 #define MAMBA_CORE_GRAPH_UTIL_HPP
 
-#include <map>
+#include <algorithm>
 #include <utility>
 #include <vector>
+#include <map>
 
 namespace mamba
 {
-    // Simplified implementation of a directed graph
-    // where a path exists between each node and the
-    // first one (you can think of it as a tree with
-    // potential cycles)
 
-    template <class T>
-    class graph
+    // A sorted vector behaving like a set.
+    template <typename... Args>
+    class vector_set : private std::vector<Args...>
     {
     public:
-        using node = T;
-        using node_id = size_t;
-        using node_list = std::vector<node>;
-        using edge_list = std::vector<node_id>;
-        using adjacency_list = std::vector<edge_list>;
-        using cycle_list = std::vector<node_list>;
-
-        const node_list& get_node_list() const;
-        const edge_list& get_edge_list(node_id id) const;
-
-        node_id add_node(const node& value);
-        node_id add_node(node&& value);
+        using Base = std::vector<Args...>;
+        using typename Base::value_type;
+
+        using Base::begin;
+        using Base::end;
+        using Base::rbegin;
+        using Base::rend;
+
+        using Base::cbegin;
+        using Base::cend;
+        using Base::crbegin;
+        using Base::crend;
+
+        using Base::clear;
+        using Base::empty;
+        using Base::size;
+
+        vector_set() = default;
+        vector_set(std::initializer_list<value_type> il);
+        template <typename InputIterator>
+        vector_set(InputIterator first, InputIterator last);
+
+        bool contains(value_type const&) const;
+
+        void insert(value_type&& value);
+        void insert(value_type const& value);
+
+    private:
+        template <typename U>
+        void insert_impl(U&& value);
+
+        template <typename... A>
+        friend bool operator==(vector_set<A...> const& lhs, vector_set<A...> const& rhs);
+    };
+
+    template <typename... Args>
+    bool operator==(vector_set<Args...> const& lhs, vector_set<Args...> const& rhs);
+
+    // Simplified implementation of a directed graph
+    template <typename Node, typename Derived>
+    class DiGraphBase
+    {
+    public:
+        using node_t = Node;
+        using node_id = std::size_t;
+        using node_list = std::vector<node_t>;
+        using node_id_list = vector_set<node_id>;
+        using adjacency_list = std::vector<node_id_list>;
+
+        node_id add_node(const node_t& value);
+        node_id add_node(node_t&& value);
         void add_edge(node_id from, node_id to);
 
+        bool empty() const;
+        std::size_t number_of_nodes() const;
+        const node_list& nodes() const;
+        node_list& nodes();
```

Comment:
This method should be removed, since an external change in the list may lead to an inconsistent state of an instance of `DiGraphBase`.

Reply:
```cpp
        node_t const& node(node_id id) const;
        node_t & node(node_id id);
```
Since changing single node data is allowed and needed.